presto-main/src/main/java/com/facebook/presto/operator/StreamingAggregationOperator.java

```diff
@@ -0,0 +1,298 @@
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(pendingInput == null);
+        requireNonNull(page, "page is null");
+
+        processInput(page);
+        updateMemoryUsage();
+    }
+
+    private void updateMemoryUsage()
+    {
+        long memorySize = pageBuilder.getRetainedSizeInBytes();
+        for (Aggregator aggregator : aggregates) {
+            memorySize += aggregator.getEstimatedSize();
+        }
+
+        if (step.isOutputPartial()) {
+            systemMemoryContext.setBytes(memorySize);
+        }
+        else {
+            userMemoryContext.setBytes(memorySize);
+        }
+    }
+
+    private void processInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        if (previousPage != null) {
+            if (!rowEqualsRow(0, previousPage, 0, page)) {
+                flushAndResetAggregates(previousPage, 0);
+                previousPage = null;
+
+                if (pageBuilder.getPositionCount() >= minOutputPageRowCount || pageBuilder.getRetainedSizeInBytes() >= minOutputPageSize.toBytes()) {
+                    pendingInput = page;
+                    hasOutput = true;
+                    return;
+                }
+            }
+            previousPage = null;
+        }
+
+        int startPosition = 0;
+        while (true) {
+            int groupEnd = findGroupEnd(startPosition, page.getPositionCount(), page);
+            for (Aggregator aggregator : aggregates) {
+                aggregator.processPage(page.getRegion(startPosition, groupEnd - startPosition));
+            }
+
+            if (groupEnd < page.getPositionCount()) {
+                flushAndResetAggregates(page, startPosition);
+
+                if (pageBuilder.getPositionCount() >= minOutputPageRowCount || pageBuilder.getRetainedSizeInBytes() >= minOutputPageSize.toBytes()) {
+                    pendingInput = page.getRegion(groupEnd, page.getPositionCount() - groupEnd);
+                    hasOutput = true;
+                    return;
+                }
+
+                startPosition = groupEnd;
+            }
+            else {
+                previousPage = page.getRegion(page.getPositionCount() - 1, 1);
+                pendingInput = null;
+                return;
+            }
+        }
+    }
+
+    private void flushAndResetAggregates(Page page, int position)
+    {
+        pageBuilder.declarePosition();
+        for (int i = 0; i < groupByTypes.size(); i++) {
+            Block block = page.getBlock(groupByChannels.get(i));
+            Type type = groupByTypes.get(i);
+            type.appendTo(block, position, pageBuilder.getBlockBuilder(i));
+        }
+        int offset = groupByTypes.size();
+        for (int i = 0; i < aggregates.size(); i++) {
+            aggregates.get(i).evaluate(pageBuilder.getBlockBuilder(offset + i));
+        }
+
+        aggregates = setupAggregates(step, accumulatorFactories);
```

Comment:
@sopel39 Karol, 

I looked at a sample implementation of the `Accumulator` generated by `AccumulatorCompiler`. (I modified `CompilerUtils#defineClass` to print out generated classes by adding `System.out.println(BytecodeUtils.dumpBytecodeTree(classDefinition));` and ran `TestArbitraryAggregation#testNullArray`. @electrum @dain Thanks to David and Dain for the pointers.)

`Accumulator`'s constructor doesn't do much. It checks arguments for `null` and instantiates an `AccumulatorState`. A `reset` method can skip `null` checks, but it still needs to create a clean instance of `AccumulatorState`. I don't think this will save us much. Perhaps, a `reset` method in `AccumulatorState` could be helpful, but that requires changing  all the existing aggregation functions. Feels like a non-starter.

Additionally, I'm thinking that `HashAggregationOperator` uses a separate instance of the `AccumulatorState` for each group anyway (except for aggregation functions which provide custom `GroupedAccumulatorState`). If that's the case, then `StreamingAggregationOperator` doesn't introduce any new inefficiency.

Finally, I updated the benchmark to test both hash-based and streaming aggregations and I'm seeing that for identical datasets streaming is faster. I'm also seeing that the performance differences between different group sizes are similar in both streaming and hash-based aggregations.

Thoughts?

```
# Run complete. Total time: 00:08:57

Benchmark                                        (operatorType)  (rowsPerGroup)  Mode  Cnt    Score    Error  Units
BenchmarkStreamingAggregationOperator.benchmark       streaming               1  avgt   30  368.381 ±  7.840  ms/op
BenchmarkStreamingAggregationOperator.benchmark       streaming              10  avgt   30   67.358 ±  3.598  ms/op
BenchmarkStreamingAggregationOperator.benchmark       streaming            1000  avgt   30   28.380 ±  0.596  ms/op
BenchmarkStreamingAggregationOperator.benchmark            hash               1  avgt   30  495.633 ± 21.028  ms/op
BenchmarkStreamingAggregationOperator.benchmark            hash              10  avgt   30  102.998 ±  3.462  ms/op
BenchmarkStreamingAggregationOperator.benchmark            hash            1000  avgt   30   53.453 ±  1.026  ms/op
```

Reply:
nice comparison against hash aggregation. I think we are fine then, streaming is faster which is good